File: verifier-cli/src/main/java/com/jetbrains/pluginverifier/verifiers/clazz/InterfacesVerifier.java
package com.jetbrains.pluginverifier.verifiers.clazz;

import com.intellij.structure.resolvers.Resolver;
import com.jetbrains.pluginverifier.location.ProblemLocation;
import com.jetbrains.pluginverifier.problems.ClassNotFoundProblem;
import com.jetbrains.pluginverifier.problems.IncompatibleClassChangeProblem;
import com.jetbrains.pluginverifier.verifiers.VContext;
import com.jetbrains.pluginverifier.verifiers.util.VerifierUtil;
import org.objectweb.asm.tree.ClassNode;

/**
 * Check that all explicitly defined interfaces exists.
 *
 * @author Dennis.Ushakov
 */
public class InterfacesVerifier implements ClassVerifier {
  public void verify(final ClassNode clazz, final Resolver resolver, final VContext ctx) {
    for (Object o : clazz.interfaces) {
      final String iface = (String)o;
      ClassNode node = VerifierUtil.findClass(resolver, iface, ctx);
      if (node == null) {
        if (!ctx.getVerifierOptions().isExternalClass(iface)) {
          ctx.registerProblem(new ClassNotFoundProblem(iface), ProblemLocation.fromClass(clazz.name));
        }
        continue;
      }
      if (!VerifierUtil.isInterface(node)) {
        ctx.registerProblem(new IncompatibleClassChangeProblem(iface, IncompatibleClassChangeProblem.Change.INTERFACE_TO_CLASS), ProblemLocation.fromClass(clazz.name));
      }
    }
  }
}

File: verifier-core/src/main/java/com/jetbrains/pluginverifier/verifiers/clazz/InterfacesVerifier.kt
package com.jetbrains.pluginverifier.verifiers.clazz

import com.jetbrains.pluginverifier.results.problems.SuperInterfaceBecameClassProblem
import com.jetbrains.pluginverifier.verifiers.VerificationContext
import com.jetbrains.pluginverifier.verifiers.resolution.ClassFile

/**
 * Check that all explicitly defined interfaces exist and are indeed interfaces (not classes).
 * If any of the classes or interfaces named as direct superinterfaces of C is not in fact an interface,
 * class loading throws an IncompatibleClassChangeError.
 */
class InterfacesVerifier : ClassVerifier {
  override fun verify(classFile: ClassFile, context: VerificationContext) {
    classFile.interfaces
        .mapNotNull { context.classResolver.resolveClassChecked(it, classFile, context) }
        .filterNot { it.isInterface }
        .forEach { context.problemRegistrar.registerProblem(SuperInterfaceBecameClassProblem(classFile.location, it.location)) }
  }
}
